aptos-move/block-executor/src/txn_last_input_output.rs

```diff
@@ -277,21 +305,13 @@ impl<T: Transaction, O: TransactionOutput<Txn = T>, E: Debug + Send + Clone>
             })
     }
 
-    pub(crate) fn resource_write_set(
+    pub(crate) fn take_resource_write_set(
         &self,
         txn_idx: TxnIndex,
-    ) -> Option<Vec<(T::Key, (T::Value, Option<Arc<MoveTypeLayout>>))>> {
-        self.outputs[txn_idx as usize]
-            .load_full()
-            .and_then(|txn_output| match &txn_output.output_status {
-                ExecutionStatus::Success(t) | ExecutionStatus::SkipRest(t) => {
-                    Some(t.resource_write_set())
-                },
-                ExecutionStatus::Abort(_)
-                | ExecutionStatus::DirectWriteSetTransactionNotCapableError
-                | ExecutionStatus::SpeculativeExecutionAbortError(_)
-                | ExecutionStatus::DelayedFieldsCodeInvariantError(_) => None,
-            })
+    ) -> Vec<(T::Key, Arc<T::Value>, Option<Arc<MoveTypeLayout>>)> {
+        self.arced_resource_writes[txn_idx as usize]
+            .swap(None)
+            .map_or(Vec::new(), |arc_vec| (*arc_vec).clone())
```

Comment:
is it fine if it is none? should we assert it is not none?

Reply:
Thx for noticing! - did with ExplicitSyncWrapper, similar to finalized groups, looks much better than way.